public/js/ClusterPlot.js

```diff
@@ -0,0 +1,212 @@
+        .attr('height', self.height + self.margin.top + self.margin.bottom)
+        .append('g')
+        .attr('transform', 'translate(' + self.margin.left + ',' + self.margin.top + ')')
+        .call(tip);
+
+    // add scatter points
+    self.svg.selectAll('circle')
+        .data(self.sampleData)
+        .enter()
+        .append('circle')
+        .attr('cx', function(d) {
+            return xScale(d.pca[0]);
+        })
+        .attr('cy', function(d) {
+            return yScale(d.pca[1]);
+        })
+        .attr('r', self.pointRadius)
+        .attr('stroke', 'white')
+        .attr('fill', function(d) {
+            var cluster_id = d.cluster_member[self.clusterMid-self.clusterMin];
+            return self.colourScale(cluster_id);
+        })
+        .attr('id', function(d) {
+            return d._id;
+        })
+        .on('mouseover', tip.show)
+        .on('mouseout', tip.hide);
+
+    // append axis
+    self.svg.append('g')
+        .attr('class', 'x axis')
+        .attr('transform', 'translate(0,' + self.height + ')')
+        .call(xAxis);
+
+    self.svg.append('g')
+        .attr('class', 'y axis')
+        .call(yAxis);
+
+    // append axis labels
+    self.svg.append('text')
+        .attr('transform', 'translate(' + self.width/2 + ' ,' + (self.height +
+        self.margin.bottom) + ')')
+        .style('text-anchor', 'middle')
+        .text('PC1');
+
+    self.svg.append('text')
+        .attr('transform', 'rotate(-90)')
+        .attr('y', -self.margin.left)
+        .attr('x', -self.height/2)
+        .attr('dy', '1em')
+        .style('text-anchor', 'middle')
+        .text('PC2');
+};
+
+/**
+ * redrawClusters: Update the scatterplot points.
+ *
+ * When a new value of k is chosen, update the colouring of the
+ * scatterplot points and their tool-tips to show which cluster they
+ * belong to.
+ *
+ * @this {redrawClusters}
+ * @param {int} newk - The number of k partitions to cluster the data into.
+ */
+ClusterPlot.prototype.redrawClusters = function(newk) {
+    var self = this;
+
+    //reset tooltips
+    var tip = d3.tip()
+        .attr('class', 'd3-tip')
+        .offset([-10, 0])
+        .html(function(d) {
+            return "<p><strong>ID: </strong>" + d._id + "</p>" +
+                "<p><strong>Gene: </strong>" + d.gene_name + "</p>" +
+                "<p><strong>Cluster: </strong>" +
+                (d.cluster_member[newk-self.clusterMin] + 1) + "</span></p>";
+        });
+
+    d3.selectAll('.d3-tip').remove();
+
+    // recolour points
+    self.svg.selectAll('circle')
+        .data(self.sampleData)
+        .attr('fill', function(d) {
+            var cluster_id = d.cluster_member[newk-self.clusterMin];
+            return self.colourScale(cluster_id);
+        })
+        .on('mouseover', tip.show)
+        .on('mouseout', tip.hide)
+        .call(tip);
+};
+
+/**
+ * destroy: Remove all child SVG elements of the plot objects containing div.
+ *
+ * @this {Lineplot}
```

Comment:
Nevermind, I see now Lineplot is your own. But is it the right `this`, here? Or just the result of some over-eager copy-pasting? ;)